test/e2e/operator_test.go

```diff
@@ -62,240 +49,185 @@ func init() {
+		panic(err)
 	}
-)
+}
 
-func TestMain(m *testing.M) {
+func initKubeClient() error {
 	kubeConfig, err := config.GetConfig()
 	if err != nil {
-		fmt.Printf("failed to get kube config: %s\n", err)
-		os.Exit(1)
+		return fmt.Errorf("failed to get kube config: %w\n", err)
 	}
 
-	kubeClient, err := client.New(kubeConfig, client.Options{})
+	kubeClient, err = client.New(kubeConfig, client.Options{})
 	if err != nil {
-		fmt.Printf("failed to create kube client: %s\n", err)
-		os.Exit(1)
+		return fmt.Errorf("failed to create kube client: %w\n", err)
 	}
-	kclient = kubeClient
+	return nil
+}
 
-	if err := kclient.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, &infraConfig); err != nil {
-		fmt.Printf("failed to get infrastructure config: %v\n", err)
-		os.Exit(1)
+func initProviderHelper() (providerTestHelper, error) {
+	var (
+		openshiftCI  bool
+		platformType string
+		err          error
+	)
+
+	if os.Getenv("OPENSHIFT_CI") != "" {
+		openshiftCI = true
+		platformType, err = getPlatformType(kubeClient)
+		if err != nil {
+			return nil, fmt.Errorf("failed to determine platform type: %w", err)
+		}
+	} else {
+		platformType = mustGetEnv("CLOUD_PROVIDER")
+	}
+
+	switch platformType {
+	case string(configv1.AWSPlatformType):
+		var (
+			awsAccessKeyID     string
+			awsSecretAccessKey string
+		)
+		if openshiftCI {
+			awsAccessKeyID, awsSecretAccessKey, err = rootAWSCredentials(kubeClient)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get AWS credentials from CCO: %w", err)
+			}
+		} else {
+			awsAccessKeyID = mustGetEnv("AWS_ACCESS_KEY_ID")
+			awsSecretAccessKey = mustGetEnv("AWS_SECRET_ACCESS_KEY")
+		}
+		return newAWSHelper(awsAccessKeyID, awsSecretAccessKey)
+	default:
+		return nil, fmt.Errorf("unsupported Provider: '%s'", platformType)
 	}
+}
 
-	platform := infraConfig.Status.PlatformStatus
-	if platform == nil {
-		fmt.Printf("platform status is missing for infrastructure %s", infraConfig.Name)
+func TestMain(m *testing.M) {
+	var (
+		err error
+	)
+	if err = initKubeClient(); err != nil {
+		fmt.Print(err)
 		os.Exit(1)
 	}
-	switch platform.Type {
-	case configv1.AWSPlatformType:
-		mySession := awsSession.Must(awsSession.NewSession())
-		dnsClient.aws = route53.New(mySession)
-	default:
-		fmt.Printf("Unsupported Provider")
+
+	if helper, err = initProviderHelper(); err != nil {
+		fmt.Print(err)
 		os.Exit(1)
 	}
 
-	/*
-		defaultParentDomainName = os.Getenv("EXTDNS_PARENT_DOMAIN")
-		if defaultParentDomainName == "" {
-			fmt.Printf("No parent domain specified. Please set the environment variable EXTDNS_PARENT_DOMAIN\n")
-			os.Exit(1)
-		}
-		defaultParentZoneID = os.Getenv("EXTDNS_PARENT_ZONEID")
-		if defaultParentZoneID == "" {
-			fmt.Printf("No parent zone ID specified. Please set the environment variable EXTDNS_PARENT_ZONEID\n")
-			os.Exit(1)
-		}
-
-		domainName = generateDomainName()
+	hostedZoneID, nameServers, err = helper.ensureHostedZone(hostedZoneDomain)
```

Comment:
If there are 2 PR running this e2e test, won't they race on the usage on the same hostedZone? Knowing also that the test service has the hard coded name, there can be false positive and negative tests.
This may need to be refactored in the followups to try to get the unique hostedzone name with the commit checksum in the name, for instance.

Reply:
Yes, this need to be done otherwise we might get too many flakes. Let's not merge this as-is.